### PR TITLE
Bingo ticket sort

### PR DESCRIPTION
**Pull request by LiamArmour**
20 files sorry :)

### CODE REVIEW

File: main-app/app/html/inGame.html [hunk 1]
```diff
@@ -13,46 +13,60 @@
...
-                <div ticket-square="0" class="bingoCell textCenter"></div>
-                <div ticket-square="1" class="bingoCell textCenter"></div>
-                <div ticket-square="2" class="bingoCell textCenter"></div>
-                <div ticket-square="3" class="bingoCell textCenter"></div>
-                <div ticket-square="4" class="bingoCell textCenter"></div>
-                <div ticket-square="5" class="bingoCell textCenter"></div>
-                <div ticket-square="6" class="bingoCell textCenter"></div>
-                <div ticket-square="7" class="bingoCell textCenter"></div>
-                <div ticket-square="8" class="bingoCell textCenter"></div>
+                <div ticket-square="0" class="textCenter"></div>
```
Comment (keithbarrow) [inline main-app/app/html/inGame.html:39]: These squares look like a good candidate for a directive in later braches (No action in this branch)

File: main-app/app/scripts/game/services/bingo-call.js [hunk 2]
```diff
@@ -1,29 +1,59 @@
...
-                    }).catch(function (data) {
-                        /* Error stub */
-                        console.log(data);
-                    });
-            };
+                makeApiCall = function (apiName, action, data) {
+                    gameProxy.callApi(apiName, action, data)
+                        .then(function (data) {
+                            me.callMessage = data;
+                            me.lastCallsDisplay.push(data.payload.call);
```
Comment (keithbarrow) [inline main-app/app/scripts/game/services/bingo-call.js]: Create a new function out of this and the next line - this adds a new call to the "recents" list and removes if necessary.

File: main-app/tests/mocks/mocks.js [hunk 3]
```diff
@@ -2,16 +2,33 @@ var mocks;
...
         $state:{
             go: function () {},
             $current: 'playerSelection'
         },
         BingoTicket:{
+            balls: [],
+            markedNumbers: [],
+            pushArray: function () {},
+        },
+        BingoCall:{
```
Comment (keithbarrow) [inline main-app/tests/mocks/mocks.js]: image

File: main-app/app/scripts/game/services/bingo-ticket.js [hunk 4]
```diff
@@ -1,41 +1,52 @@
...
-                var endOfString = 30;
-                for(startOfString; startOfString < ticket.length; startOfString += 30){
-                    tickets.push(ticket.slice(startOfString, endOfString));
-                    endOfString += 30;
-                }
-
-                for(ticket in tickets){
-                    splitUpTheTicket(tickets[ticket]);
+            me.pushArray = function(card) {
+                var cardss = "054963758028345266770611596982054963758028345266770611596982",
```
Comment (keithbarrow) [inline main-app/app/scripts/game/services/bingo-ticket.js]: Really? Hardcoded? 
Also `card` not `cards`
Also this is almost useless from a programmatic point of view - a whacking great string in place of what is really a 10x4 grid.

File: main-app/app/scripts/game/services/bingo-ticket.js [hunk 5]
```diff
@@ -1,41 +1,52 @@
...
+                    i;
+                for(i = 0; i < 27; i++){
+                    me.markedNumbers[i] = "bingoCell";
```
Comment (keithbarrow) [inline main-app/app/scripts/game/services/bingo-ticket.js:15]: This makes no sense - why are you filling 27 numbers with 'bingoCell'?

File: main-app/app/scripts/game/services/bingo-ticket.js [hunk 6]
```diff
@@ -1,41 +1,52 @@
...
-                var endOfString = 2;
-                for(startOfString; startOfString < ticket.length; startOfString += 2){
-                    me.balls.push(ticket.slice(startOfString, endOfString));
-                    endOfString += 2;
-                }
-                console.log(me.balls);
-            };
-
-
+            var generateTheTicket = function(ticket) {
```
Comment (keithbarrow) [inline main-app/app/scripts/game/services/bingo-ticket.js:20]: Not going to continue to reviewing this file. 
Need to think more in  terms of objects:
1. A card has a an array of 4 lines. If all the lines have 0 to go, they you've won a house prize.
2. A line is made up of an array of 9 squares. You can calculate how many to go by adding up which squares are "to go".
3.  A square has a value or is null. It is "to go" if it is not null and the number hasn't been called.

Refactoring this way should make the other tasks easer: you just add a function callMade(ball) and the ticket updates itself.

File: main-app/app/scripts/game/services/end-of-game-service.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+            me.checkForWinner = function(response) {
+                if (response.message === "Line") {
+                    me.checkForLine(response);
+                }
+                if (response.message === "Winner") {
+                    me.checkForHouse(response);
+                }
+            };
+
+            me.checkForLine = function(response) {
```
Comment (keithbarrow) [inline main-app/app/scripts/game/services/end-of-game-service.js]: This isn't checking for a line - it is alerting the user a line has been won.

File: main-app/app/scripts/game/services/end-of-game-service.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+                alert('--Line Prize\nWinner: ' + response.payload.winnerInfo.linewinnername + '\nAmount Won: ' + response.payload.winnerInfo.lineprize);
+            };
+
+            me.checkForHouse = function(response) {
```
Comment (keithbarrow) [inline main-app/app/scripts/game/services/end-of-game-service.js]: This isn't checking for a house prize - it is alerting the user.

File: main-app/app/scripts/game/services/end-of-game-service.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+                if (response.message === "Line") {
+                    me.checkForLine(response);
+                }
+                if (response.message === "Winner") {
+                    me.checkForHouse(response);
+                }
+            };
+
+            me.checkForLine = function(response) {
+                alert('--Line Prize\nWinner: ' + response.payload.winnerInfo.linewinnername + '\nAmount Won: ' + response.payload.winnerInfo.lineprize);
```
Comment (keithbarrow) [inline main-app/app/scripts/game/services/end-of-game-service.js:18]: Put a todo on this: alert is not a viable solution for anything.

File: main-app/app/scripts/game/services/end-of-game-service.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+            };
+
+            me.checkForHouse = function(response) {
+                alert('--House Prize\nWinner: ' + response.payload.winnerInfo.housewinnername + '\nAmount Won: ' + response.payload.winnerInfo.houseprize + '\n--Line Prize\nWinner: ' + response.payload.winnerInfo.linewinnername + '\nAmount Won: ' + response.payload.winnerInfo.lineprize);
```
Comment (keithbarrow) [inline main-app/app/scripts/game/services/end-of-game-service.js:22]: Put a todo on this: alert is not a viable solution for anything.

File: main-app/app/scripts/game/services/end-of-game-service.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+                me.gameEnded();
+            };
+
+            me.gameEnded = function() {
```
Comment (keithbarrow) [inline main-app/app/scripts/game/services/end-of-game-service.js:26]: This is part of the bingo-calling functionality - the fact `me.gameloop` isn't part of this service should perhaps have tipped you off.

File: main-app/app/scripts/game/services/game-api.js [hunk 12]
```diff
@@ -1,23 +1,18 @@
...
-                },
-
-            callApi = function(apiName, action, data, token){
-                    gameProxy.callApi(apiName, action, data, token)
+                callApi = function (apiName, action, data) {
+                    gameProxy.callApi(apiName, action, data)
                         .then(function (data) {
-                            me.returnedMessage = data;
+                            coreApiConverter.convert(data);
                             $state.go(data.message);
```
Comment (keithbarrow) [inline main-app/app/scripts/game/services/game-api.js:19]: The proxy should't be managing state changes - that is UI work.

File: main-app/app/scripts/game/services/game-api.js [hunk 13]
```diff
@@ -1,23 +1,18 @@
...
-                            console.log(data);
-                            if(data.message == "TicketBought"){
+                            if (data.message == "TicketBought") {
                                 bingoTicket.pushArray(data.payload.card);
```
Comment (keithbarrow) [inline main-app/app/scripts/game/services/game-api.js:22]: Provide a callback to update the ticket, or wrap this all up with a promise.

**Comment by keithbarrow** [comment]
Not reviewing the unit tests - these are likely to substantially change

File: main-app/app/scripts/game/services/bingo-ticket.js (hunk 5 above)
Comment (LiamArmour) [inline main-app/app/scripts/game/services/bingo-ticket.js:15]: I've explained this one, i will refactor this later in a new branch

File: main-app/app/scripts/game/services/bingo-ticket.js (hunk 6 above)
Comment (LiamArmour) [inline main-app/app/scripts/game/services/bingo-ticket.js:20]: 1. A card has a an array of 4 lines, this is incorrect... it only has 3 :)

**Pull request merged**